crates/propolis-api-types/src/lib.rs

```diff
@@ -36,6 +36,10 @@ pub struct InstancePathParams {
 pub struct InstanceMetadata {
     pub silo_id: Uuid,
     pub project_id: Uuid,
+    pub sled_id: Uuid,
+    pub sled_serial: String,
+    pub sled_revision: u32,
+    pub sled_model: String,
```

Comment:
Should we add a (phd?) test for checking that these values are updated when an instance is migrated?

Reply:
Sure that seems good. I've not written any PHD tests, so might take me a bit.

Reply:
You'll probably want to start by looking at the [existing PHD tests for live migration](https://github.com/oxidecomputer/propolis/blob/master/phd-tests/tests/src/migrate.rs). Hopefully, the test-support library should be *pretty* straightforward, but do let me know if you have questions!

Reply:
I added a test in 4d5453a. It feels a bit contrived, since we're mostly testing the internal `TestVm`'s handling of instance metadata, so I'm on the fence about it. Happy to keep or remove, depending on what others think.

Reply:
With the caveat that I haven't looked at the non-PHD parts of the PR very carefully yet: I think the test would feel less contrived if the metadata were part of `struct VmSpec` instead of being randomly generated in `start_local_vm`. I'd arrange that like this:

- add an `InstanceMetadata` to `struct VmSpec`
- have `VmConfig::vm_spec` generate random IDs (right now this is in `start_local_vm`)
- add a `VmSpec` function that generates new sled IDs/revisions (in your change this is happening in `instance_ensure_internal`); call this function from `Framework::spawn_successor_vm` (just like the call to `refresh_crucible_backends`)

Then most of the changes to `struct TestVm` can go away, and your new test would look something like this:

```rust
async fn migration_ensures_instance_metadata(ctx: &Framework) {
    // Create a source instance, and fetch the instance metadata its metrics are
    // generated with.
    let mut source = ctx
        .spawn_default_vm("migration_ensures_instance_metadata_source")
        .await?;
    let mut target = ctx
        .spawn_successor_vm(
            "migration_ensures_instance_metadata_target",
            &source,
            None,
        )
        .await?;

    source.launch().await?;
    source.wait_to_boot().await?;
    let source_expected = source.vm_spec().metadata;
    let source_metadata = source.get_spec().await?.properties.metadata;
    assert_eq!(source_metadata, source_expected);

    // Migrate the instance to a new server, and refetch the metadata.
    target
        .migrate_from(&source, Uuid::new_v4(), MigrationTimeout::default())
        .await?;
    let target_expected = target.vm_spec().metadata;
    let target_metadata = target.get_spec().await?.properties.metadata;
    assert_eq!(target_metadata, target_expected);

    // gjc: I might not keep the "project and silo IDs shouldn't change"
    // requirement; the control plane has to uphold that, not Propolis. You
    // could have a unit test for `VmSpec::refresh_sled_info` that makes sure
    // this invariant is upheld, though.
}
```

WDYT? (Does this suggestion make any sense in view of what's going on in the rest of the PR? It very well may not, in which case you should feel free to ignore it :))

Reply:
Thanks for the suggestion @gjcolombo. That does seem slightly less contrived, in that at least the manipulation of the source and target metadata is contained and mostly transparent to the tests. I've made those updates in 7981b96, with a few minor visibility modifications to glue it all together. Thanks!